### PR TITLE
[Codegen][LLVMGPU] Set global read layouts at linalg level

### PR DESCRIPTION
**Pull request by Groverkss**
- Operand promotion is now done the same way as TileAndFuse pipeline, by reading promote_operands config from lowering_config.
- Moves global read layout setting to LLVMGPUConfigureTensorLayouts, from LLVMGPUConfigureVectorLayouts pass, anchoring based on lowering config.

These changes by side effect allow setting layouts on gathers in VectorDistribute pipeline.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUConfigureTensorLayouts.cpp [hunk 1]
```diff
@@ -292,21 +334,122 @@ setAttentionMatmulAnchor(IREE::GPU::MMAScheduleAttr schedule,
...
+
+  SmallVector<int64_t> threadTile(opRank, 1);
+  SmallVector<int64_t> threadStrides(opRank, 0);
+
+  int64_t residualThreads = ShapedType::getNumElements(workgroupSize);
+  int64_t currStride = 1;
+
+  for (auto [tile, stride, size] :
+       llvm::reverse(llvm::zip(threadTile, threadStrides, opShape))) {
+    int64_t threadTile;
```
Comment (raikonenfnu) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUConfigureTensorLayouts.cpp]: probably good to change the name, seems like we have two threadTile

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUConfigureTensorLayouts.cpp [hunk 2]
```diff
@@ -292,21 +334,122 @@ setAttentionMatmulAnchor(IREE::GPU::MMAScheduleAttr schedule,
...
+    if (residualThreads % size == 0) {
+      threadTile = size;
+    } else if (size % residualThreads == 0) {
+      threadTile = residualThreads;
+    } else {
```
Comment (raikonenfnu) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUConfigureTensorLayouts.cpp:412]: Looks/feels kind of similar to the code we have in ConfigureVectorLayout? are we planning to remove the code there?

Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUConfigureTensorLayouts.cpp:412]: Yes, i updated the patch to remove it.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUConfigureTensorLayouts.cpp (hunk 1 above)
Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUConfigureTensorLayouts.cpp]: oh true, will fix

**Review by MaheshRavishankar — APPROVED** [review]
I dont have enough handle on the details, but overall looks good to me. OK to land, unless someone else has an objection.

**Review by raikonenfnu — APPROVED** [review]
LGTM!

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUConfigureTensorLayouts.cpp [hunk 3]
```diff
@@ -137,30 +165,38 @@ static LogicalResult setConvolutionAnchor(IREE::GPU::MMAScheduleAttr schedule,
   Value rhs = conv->getOperand(1);
   Value acc = conv->getOperand(2);
 
-  // Set layouts for lhs, rhs and acc.
+  // eet layouts for lhs, rhs and acc.
```
Comment (raikonenfnu) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUConfigureTensorLayouts.cpp]: NIT: Typo to "Set"

Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUConfigureTensorLayouts.cpp]: done :D

**Pull request merged**